chainer/reporter.py

```diff
@@ -0,0 +1,325 @@
+import collections
+import contextlib
+
+import numpy
+import six
+
+from chainer import cuda
+from chainer import variable
+
+
+class Reporter(object):
+
+    """Object to which observed values are reported.
+
+    Reporter is used to collect values that users want to watch. The reporter
+    object holds a mapping from value names to the actually observed values.
+
+    When a value is passed to the reporter, an observer object can be
+    optionally attached. In this case, the name of the observer is added as the
+    prefix of the value name. The observer name should be registered
+    beforehand.
+
+    See the following example::
+
+       >>> from chainer import Reporter, report, report_scope
+       >>>
+       >>> reporter = Reporter()
+       >>> observer = object()  # it can be an arbitrary (reference) object
+       >>> reporter.add_observer('my_observer:', observer)
+       >>> observation = {}
+       >>> with reporter.scope(observation):
+       ...     reporter.report({'x': 1}, observer)
+       ...
+       >>> observation
+       {'my_observer:x': 1}
+
+    There are also a global API to add values::
+
+       >>> observation = {}
+       >>> with report_scope(observation):
+       ...     report({'x': 1}, observer)
+
+    The most important application of Reporter is to report observed values
+    from each link or chain in the training and validation procedures.
+    :class:`~chainer.training.Trainer` and some extensions prepare their own
+    Reporter object with the hierarchy of the target link registered as
+    observers. We can use :func:`report` function inside any links and chains
+    to report the observed values (e.g., training loss, accuracy, activation
+    statistics, etc.).
+
+    Attributes:
+        observation: Dictionary of observed values.
+
+    """
+    def __init__(self):
+        self._observer_names = {}
+        self.observation = {}
+
+    def __enter__(self):
+        """Makes this reporter object current."""
+        _reporters.append(self)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Recovers the previous reporter object to the current."""
+        _reporters.pop()
+
+    @contextlib.contextmanager
+    def scope(self, observation):
```

Comment:
The term `scope` is used without defining its meaning. I guess briefly explaining the concept of `scope` somewhere (the description of this function or Reporter class itself?) eases the understanding of Reporter.